File: Scala/AoC2024/src/main/scala/Day01.scala
import AoCLib.AoCUtil

import scala.collection.mutable.ArrayBuffer

class Day01(day: Int, name: String) extends AoCLib.Solution(day, name):
  override def solve(test: Boolean, index: Int): Unit =
    val input = AoCUtil.readInput(AoCUtil.fileName(day, test))
      .map(line =>
        val pattern = """(\d+)\s+(\d+)""".r
        line match
          case pattern(d1, d2) => List(d1.toInt, d2.toInt)
          case _ => throw IllegalArgumentException(s"Could not parse $line"))
    var pivot = AoCUtil.pivotMatrix(input)

    solvePartOne(pivot)
    solvePartTwo(pivot)

  private def solvePartOne(input: ArrayBuffer[ArrayBuffer[Int]]): Unit =
    val sorted = input.map(buf => buf.sorted)

    val totalDistance = (sorted(0) zip sorted(1))
      .map(pair => math.abs(pair._1 - pair._2))
      .sum
    println(s"Part One: the total distance is $totalDistance")

  private def solvePartTwo(input: ArrayBuffer[ArrayBuffer[Int]]): Unit =
    val col2counts = input(1).groupMapReduce(_.toInt)(_ => 1)(_ + _)

    val totalSimilarity = input.head
      .map(value =>
        if col2counts.contains(value) then
          (value * col2counts(value)) else 0)
      .sum

    println(s"Part Two: the total similarity is $totalSimilarity")

File: F#/AoC2024/Day01.fs
[<AutoOpen>]
module Day01

open AoC.Util
open System.Text.RegularExpressions

let parseInput input =
    let rx = Regex(@"(\d+)\s+(\d+)")
    let check e = 
        let m = rx.Match(e)
        [int m.Groups[1].Value; int m.Groups[2].Value]
    
    input |> List.map check

let solvePartOne input =
    let sorted = input |> List.map List.sort
    
    let totalDistance =
        List.zip sorted[0] sorted[1]
        |> List.map (fun (a,b) -> abs(a - b))
        |> List.sum
    totalDistance

let solvePartTwo (input: int list list) = 
    let col2Counts = frequencyMap input[1]
    let totalSimilarity =
        input[0]
        |> List.map (fun v ->
            if col2Counts.ContainsKey v then
                v * col2Counts.Item v
            else
                0
        )
        |> List.sum
    totalSimilarity

let solveDay01 isTest: Unit =
    let day = 1
    let puzzleName = "Historian Hysteria"
    printfn $"Day {day}: {puzzleName}"
    let inputName = inputFileName day isTest
    let input = readInput inputName true
    
    let pairs = parseInput input
    let pivot = AoC.Util.pivotMatrix pairs
    
    let solution1 = solvePartOne pivot
    printfn $"The total distance is {solution1}"

    let solution2 = solvePartTwo pivot
    printfn $"The total similarity is {solution2}"

    printfn "All done."
